python/ray/signature.py

```diff
@@ -131,92 +88,125 @@ def extract_signature(func, ignore_first=False):
-            keyword_names.add(arg_name)
-
-    return FunctionSignature(arg_names, arg_defaults, arg_is_positionals,
-                             keyword_names, func.__name__)
+        signature_parameters = signature_parameters[1:]
 
+    return _scrub_parameters(signature_parameters)
 
-def extend_args(function_signature, args, kwargs):
-    """Extend the arguments that were passed into a function.
 
-    This extends the arguments that were passed into a function with the
-    default arguments provided in the function definition.
+def validate_args(signature_parameters, args, kwargs):
+    """Checks the function signature and validates the provided arguments.
 
     Args:
         function_signature: The function signature of the function being
             called.
         args: The non-keyword arguments passed into the function.
         kwargs: The keyword arguments passed into the function.
 
-    Returns:
-        An extended list of arguments to pass into the function.
-
     Raises:
         Exception: An exception may be raised if the function cannot be called
             with these arguments.
     """
-    arg_names = function_signature.arg_names
-    arg_defaults = function_signature.arg_defaults
-    arg_is_positionals = function_signature.arg_is_positionals
-    keyword_names = function_signature.keyword_names
-    function_name = function_signature.function_name
-
-    args = list(args)
-
-    for keyword_name in kwargs:
-        if keyword_name not in keyword_names:
-            raise Exception("The name '{}' is not a valid keyword argument "
-                            "for the function '{}'.".format(
-                                keyword_name, function_name))
-
-    # Fill in the remaining arguments.
-    for skipped_name in arg_names[0:len(args)]:
-        if skipped_name in kwargs:
-            raise Exception("Positional and keyword value provided for the "
-                            "argument '{}' for the function '{}'".format(
-                                keyword_name, function_name))
-
-    zipped_info = zip(arg_names, arg_defaults, arg_is_positionals)
-    zipped_info = list(zipped_info)[len(args):]
-    for keyword_name, default_value, is_positional in zipped_info:
-        if keyword_name in kwargs:
-            args.append(kwargs[keyword_name])
+    restored = _restore_parameters(signature_parameters)
+    reconstructed_signature = funcsigs.Signature(parameters=restored)
+    try:
+        reconstructed_signature.bind(*args, **kwargs)
+    except TypeError as exc:
+        raise TypeError(str(exc))
+
+
+def flatten_args(args, kwargs):
+    """Generates a serializable format for the arguments.
+
+    Args:
+        args: The non-keyword arguments passed into the function.
+        kwargs: The keyword arguments passed into the function.
+
+    Returns:
+        List of args and kwargs. Non-keyword arguments are prefixed
+            by internal enum DUMMY_TYPE.
```

Comment:
I assume there's a good reason for having the `DUMMY_TYPE` but it's not immediately obvious to me from the code. Please add a comment explaining.

Reply:
done